<!DOCTYPE html>
<html lang="en">
    
    

    <head>
    <link href="https://gmpg.org/xfn/11" rel="profile">
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta http-equiv="Cache-Control" content="public" />
<!-- Enable responsiveness on mobile devices -->
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">
<meta name="generator" content="Hugo 0.46" />

    
    
    

<title>Microsoft SQL Server R Services - Internals XX • Niels Berglund</title>
<meta name="description" content="nielsb&#39;s blog :: technology musings about coding and data. Some topics: .NET, SQL Server, Data Science, R, Windows Azure and a lot more">
<meta name="keywords" content="sql server, c#, distributed computing, data science, microsoft r server, microsoft machine learning server, data science, sql server r services, sql server machine learning services, kafka, flink">
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Microsoft SQL Server R Services - Internals XX"/>
<meta name="twitter:description" content="A drill down, using WinDbg, WireShark, Process Monitor, into how error messages in the external engine is handled by SQL Server."/>

<meta property="og:title" content="Microsoft SQL Server R Services - Internals XX" />
<meta property="og:description" content="A drill down, using WinDbg, WireShark, Process Monitor, into how error messages in the external engine is handled by SQL Server." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/2018/02/02/microsoft-sql-server-r-services---internals-xx/" />



<meta property="article:published_time" content="2018-02-02T19:48:22&#43;02:00"/>

<meta property="article:modified_time" content="2018-02-02T19:48:22&#43;02:00"/>












    

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/monokai.min.css">



<link rel="stylesheet" href="/css/hyde-hyde.css">
<link rel="stylesheet" href="/css/print.min.css" media="print">

    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
    <!-- Icons -->
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/apple-touch-icon-144-precomposed.png">
    <link rel="shortcut icon" href="/favicon.png">
    

</head>


    <body >
        
<div class="sidebar">
  <div class="container">
    <div class="sidebar-about">
      <span class="site__title">
        <a href="">Niels Berglund</a>
      </span>
      
      <p class="site__description">
         Technology musings about coding and data. Some topics: .NET, SQL Server, Data Science, R, Windows Azure and a lot more. 
      </p>
      <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>

<ins class="adsbygoogle"
     style="display:inline-block;width:300px;height:250px"
     data-ad-client="ca-pub-3005153158271538"
     data-ad-slot="7704601332"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>


      <div>
	<ul class="sidebar-nav">
		 
		 
			 
				<li>
					<a href="/">
						<span>Home</span>
					</a>
				</li>
			 
		 
			 
				<li>
					<a href="/posts/">
						<span>Post Archive</span>
					</a>
				</li>
			 
		 
			 
				<li>
					<a href="/series/">
						<span>Blog Post Series</span>
					</a>
				</li>
			 
		 
			 
				<li>
					<a href="/categories/">
						<span>Categories</span>
					</a>
				</li>
			 
		 
			 
				<li>
					<a href="/tags/">
						<span>Tags</span>
					</a>
				</li>
			 
		 
			 
				<li>
					<a href="/about/">
						<span>About</span>
					</a>
				</li>
			 
		 
			 
				<li>
					<a href="/disclaimer/">
						<span>Disclaimer</span>
					</a>
				</li>
			 
		
		</li>
	</ul>
</div>

    
      <img src="/images/MVP_Logo_large.png"/>

    </div>

      
    
    <p>
      <section class="social">
	<h3 style="color:#ffffff">Follow Me:</h3>
	<a href="http://feeds.feedburner.com/manageddata/"><i class="fas fa-rss"></i></a>
	
	&nbsp;<a href="https://twitter.com/nielsberglund"><i class="fab fa-twitter fa-lg" aria-hidden="true"></i></a>
	
	
	
	&nbsp;<a href="https://github.com/nberglund"><i class="fab fa-github fa-lg" aria-hidden="true"></i></a>
	
	
	
	
	
	&nbsp;<a href="https://linkedin.com/in/niels-berglund-0122593"><i class="fab fa-linkedin fa-lg" aria-hidden="true"></i></a>
	
	
	&nbsp;<a href="https://stackoverflow.com/users/7656880"><i class="fab fa-stack-overflow fa-lg" aria-hidden="true"></i></a>
	
	
	
	
	
</section>

    </p>
    <p class="copyright">
      &copy; 2019 nielsb.
      <a href="https://creativecommons.org/licenses/by-sa/4.0">Some Rights Reserved</a>.
      
    </p>
  </div>
  <div>
  </div>
</div>

        <div class="content container">
            
    <article>
  <header>
    <h1>Microsoft SQL Server R Services - Internals XX</h1>
     
    
<div class="post__meta">
    
    
      <i class="fas fa-calendar-alt"></i> Feb 2, 2018
    
    
    
      
      
          in
          
          
              <a class="post__category" href="/categories/data-science">DATA SCIENCE</a>
              •
          
              <a class="post__category" href="/categories/microsoft-r-server">MICROSOFT R SERVER</a>
              •
          
              <a class="post__category" href="/categories/sql-server">SQL SERVER</a>
              •
          
              <a class="post__category" href="/categories/sql-server-machine-learning-services">SQL SERVER MACHINE LEARNING SERVICES</a>
              •
          
              <a class="post__category" href="/categories/sql-server-r-services">SQL SERVER R SERVICES</a>
              •
          
              <a class="post__category" href="/categories/sql-server-r-services-series">SQL SERVER R SERVICES SERIES</a>
              
          
      
    
    
    
      
      
          <br/>
           <i class="fas fa-tags"></i>
          
          <a class="post__tag" href="/tags/sql-server-r-services">sql server r services</a>
           
      
          <a class="post__tag" href="/tags/sql-server-machine-learning-services">sql server machine learning services</a>
           
      
          <a class="post__tag" href="/tags/r">r</a>
           
      
          <a class="post__tag" href="/tags/windbg">windbg</a>
           
      
          <a class="post__tag" href="/tags/launchpad">launchpad</a>
           
      
          <a class="post__tag" href="/tags/rterm.exe">rterm.exe</a>
           
      
          <a class="post__tag" href="/tags/bxl">bxl</a>
           
      
          <a class="post__tag" href="/tags/process-monitor">process monitor</a>
           
      
          <a class="post__tag" href="/tags/sqlsatellite">sqlsatellite</a>
           
      
          <a class="post__tag" href="/tags/wireshark">wireshark</a>
          
      
    
    
    <br/>
    <i class="fas fa-clock"></i> 24 min read
</div>


    <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>

<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-3005153158271538"
     data-ad-slot="6668073777"
     data-ad-format="auto"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
  </header>
  <div class="post">
    <p>This is the 21:th post in a series about <strong>Microsoft SQL Server R Services</strong>, (it is now called <strong>Microsoft SQL Server Machine Learning Services</strong>), and the 20:th post that drills down into the internal of how it works.</p>

<p>Wow, 20 posts about internals - when I started this <strong>Microsoft SQL Server R Services</strong> series, I thought I might achieve five or six posts. I would never have - in my wildest fantasies - thought I would reach twenty!</p>

<p>To see other posts (including this) in the series, go to <a href="/sql_server_2k16_r_services"><strong>SQL Server R Services</strong></a>.</p>

<p>We are however now coming closer to the end of this whole series, and this post is the last for <strong>Internals</strong>. After this post, I foresee one, possibly two, more posts about other things related to <strong>Microsoft SQL Server R Services</strong>  - and then we are done!</p>

<p></p>

<p>In the last post: <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a> we discussed how &ldquo;print&rdquo; statements and acks came back to SQL Server from the external engine. When I started writing that post, I intended to cover error messages as well, but as I was investigating what happens, I realised it was more to error messages than what I initially thought. So in this post, we talk about how SQL Server and the launchpad service handle errors.</p>

<h2 id="recap">Recap</h2>

<p>Before we go into the details about errors, let us do a recap of the <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a> post, as there are many similarities between the handling of error messages and &ldquo;print&rdquo;/ack messages are.</p>

<p>The easiest way to recap is to use the two figures we had in <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a> which illustrated the flow of the launchpad service as well as SQL Server.</p>

<h4 id="launchpad-service">Launchpad Service</h4>

<p>This is the flow (on a high level) in the launchpad service:</p>

<p><img src="/images/posts/sql_r_services_19_launchpad_flow.png" alt="" /></p>

<p><strong>Figure 1:</strong> <em>Launchpad Flow</em></p>

<p>In <em>Figure 1</em> we see the main calls of leading up to and including when messages come back from the external engine. The numbered arrows show some of the calls and the communication in the launchpad service:</p>

<ol>
<li>During startup, an IOCP &ldquo;listener&rdquo; is set up.</li>
<li>Creates supporting directories and process, and initiates the connection to the external engine.</li>
<li>A call comes back from the external engine when work finishes.</li>
<li>The call causes the &ldquo;listener&rdquo; to call <code>Read...</code> routines.</li>
<li><code>LaunchServTask</code> is signalled.</li>
<li>Execution in <code>LaunchServTask</code> continues into launchpad!CLaunchContext::Cleanup.</li>
<li>A call to  <code>launchpad!CSatelliteCargoContext::SendLogMessage</code> happens if a &ldquo;print&rdquo; message has been passed back from the external engine, eventually <code>launchpad!CSatelliteCargoContext::SendLogMessage</code> is called.</li>
<li><code>launchpad!CSatelliteCargoContext::SendLogMessage</code> sends the message to SQL Server.</li>
<li><code>LaunchServTask</code> continues into <code>launchpad!CSQLSatelliteCommunication::SendAckMessage</code>, which sends an ack that the &ldquo;job is done&rdquo; to SQL Server.</li>
</ol>

<h4 id="sql-server">SQL Server</h4>

<p>This is the high level flow in SQL Server:</p>

<p><img src="/images/posts/sql_r_services_19_sql_flow.png" alt="" /></p>

<p><strong>Figure 2:</strong> <em>SQL Server Flow</em></p>

<p>The numbered arrows in <em>Figure 2</em> illustrates:</p>

<ol>
<li>During the launch of the launchpad service, an IOCP &ldquo;listener&rdquo; is set up.</li>
<li>Message sent to the launchpad service.</li>
<li>After the call above, the SqlSatellite opens a TCP/IP connection to SQL Server.</li>
<li>SQL Server sends authentication packets and script packet to the SqlSatellite.</li>
<li>If there is data for <code>@input_data_1</code>, it is sent, together with two end packets.</li>
<li>Packets are coming back to SQL Server containing meta-data and the actual data.</li>
<li><code>PullRow</code> is being signalled, and processing of meta-data as well as result rows continues.</li>
<li>The launchpad service sends &ldquo;print&rdquo;/ack messages to SQL Server. In the figure, they appear as one packet, but there is one message for &ldquo;print&rdquo; and one for ack.</li>
<li><code>ErsqExecuteQuery</code> is signalled.</li>
<li>Execution continues by calling into <code>sqllang!CXStmtQuery::ShutdownNormal</code>, which calls into <code>FinalCleanup,</code>ShutdownService, and a call to <code>sqllang!CSQLSatelliteCommunication::WaitForAckMessage</code> is made followed by <code>sqllang!CSQLSatelliteConnection::GetNextMessage</code>.</li>
<li>If a &ldquo;print&rdquo; message has is sent to SQL Server, <code>ReadPayload</code> is called and then <code>sqllang!CUDXR_ExternalScript::HandleLogMessage</code>.  The code loops on <code>sqllang!CSQLSatelliteCommunication::WaitForAckMessage</code> if there is no &ldquo;print&rdquo; message.</li>
<li>SQL Server loops on <code>sqllang!CSQLSatelliteCommunication::WaitForAckMessage</code> and waits for the ack message,</li>
<li>After receiving the ack message, SQL Server sends relevant messages to the calling code.<br /></li>
</ol>

<h2 id="housekeeping">Housekeeping</h2>

<p>Before we &ldquo;dive&rdquo; into today&rsquo;s topics let us look at the code and the tools we use today. This section is here for those of who want to follow along in what we are doing in the posts.</p>

<h4 id="helper-tools">Helper Tools</h4>

<p>To help us figure out the things we want, we use <em>Process Monitor</em>, <em>WinDbg</em> and <em>WireShark</em>:</p>

<ul>
<li><em>Process Monitor</em>, is used it to filter TCP traffic. If you want to refresh your memory about how to do it, we covered that in <a href="/2017/08/29/microsoft-sql-server-r-services---internals-x/">Internals - X</a> as well as in <a href="/2017/11/25/microsoft-sql-server-r-services---internals-xiv/">Internals - XIV</a>.</li>
<li>In the last few posts I have used <em>WinDbg preview</em> as my debugger, but in this post, I am reverting to <em>WinDbg</em> &ldquo;classic&rdquo;. If you need help with setting it up, I covered that in <a href="/2017/03/18/microsoft-sql-server-r-services---internals-i/">Internals - I</a>.</li>
<li>I use <em>WireShark</em> for network packet sniffing. If you want a refresher about <em>WireShark</em>, we covered the setup and so forth in <a href="/2017/08/29/microsoft-sql-server-r-services---internals-x/">Internals - X</a>.</li>
</ul>

<h4 id="code">Code</h4>

<p>In this post we won&rsquo;t need a specific database, we&rsquo;ll just some basic code:</p>

<pre><code class="language-sql">EXEC sp_execute_external_script
                 @language = N'R' ,
                 @script = N'Sys.sleep(10)
                             d &lt;- 42';
</code></pre>

<p><strong>Code Snippet 1:</strong> <em>Base Code to Execute</em></p>

<p>The code in <em>Code Snippet 1</em> is the code we use for &ldquo;base-lining&rdquo;, e.g. what it looks like when everything works. The <code>Sys.sleep</code> is in the script to make it easier for us to see what happens, and in what order.</p>

<p>We also use variants of that code to create errors:</p>

<pre><code class="language-sql">EXEC sp_execute_external_script
                 @language = N'R' ,
                 @script = N'Sys.sleep(10)
                             d &lt;- 42
                             #the function below does not exist
                             nonExistFunc(d)';
</code></pre>

<p><strong>Code Snippet 2:</strong> <em>Script with Missing Function</em></p>

<pre><code class="language-sql">EXEC sp_execute_external_script
                 @language = N'R' ,
                 @script = N'Sys.sleep(1)
                             d &lt;- 42',
                 @input_data_1 = N'SELECT 1/ 0';
</code></pre>

<p><strong>Code Snippet 3:</strong> <em>Error When Selecting Data</em></p>

<p>We talk more about the two code snippets below.</p>

<h2 id="error">Error</h2>

<p>When we discuss error messages, there are two main types of errors, and the code above exemplifies that:</p>

<ul>
<li>Errors generated from the external engine script - <em>Code Snippet 2</em>.</li>
<li>Errors raised when generating the input data-set (<code>@input_data_1</code>) - <em>Code Snippet 3</em>.</li>
</ul>

<p>So, let us look at what happens when we execute the code in <em>Code Snippet 2</em> and <em>Code Snippet 3</em>.</p>

<p>This is the output from <em>Code Snippet 2</em>:</p>

<p><img src="/images/posts/sql_r_services_19_error1.png" alt="" /></p>

<p><strong>Figure 3:</strong> <em>R Script Error</em></p>

<p>The output from executing the code in <em>Code Snippet 3</em>:</p>

<p><img src="/images/posts/sql_r_services_20_error2.png" alt="" /></p>

<p><strong>Figure 4:</strong> <em>SQL Execution Error</em></p>

<p>When looking at both <em>Figure 3</em> and <em>Figure 4</em>, we see that there are SQL Server error numbers assigned to the errors. However, compare the errors in <em>Figure 3</em>, and <em>Figure 4</em>. In <em>Figure 3</em> there are two errors; <code>39004</code> and <code>39019</code> - and in the error message they both refer to the external script. In <em>Figure 4</em> however, the error message is the usual message about division by zero, and no referral to external script whatsoever. There is though some text about SqlSatellite.</p>

<p>So it seems that for specific external script errors, SQL Server knows about them, but others not.</p>

<h2 id="known-sql-errors">Known SQL Errors</h2>

<p>We start by looking at, what I call, known SQL errors - like we see in <em>Figure 3</em>. Just for fun, let us execute following code:</p>

<pre><code class="language-sql">SELECT message_id, text FROM sys.messages
WHERE message_id BETWEEN 39000 AND 39999
AND language_id = 1033;
</code></pre>

<p><strong>Code Snippet 4:</strong> <em>Selecting Error Messages</em></p>

<p>On my SQL Server 2016 box the statement in <em>Code Snippet 4</em> returns 27 rows, and when browsing through the result I realise that most of them are related to the external script or the external engine:</p>

<p><img src="/images/posts/sql_r_services_20_error_msgs.png" alt="" /></p>

<p><strong>Figure 5:</strong> <em>SQL Server Error Messages</em></p>

<p>In <em>Figure 5</em> wee see some of the error codes related to the external script/engine - so the question is how SQL Server knows that, what originates from the external engine, is an error?</p>

<p>We know that error(s) are raised in SQL Server when something goes wrong in the external engine (see <em>Figure 3</em>), but what happens in the launchpad service? So, let us do this:</p>

<ul>
<li>Start up two instances of <em>WinDbg</em> as admin.</li>
<li>Attach <em>WinDbg</em> to the SQL Server process (<code>sqlservr.exe</code>), and the launchpad process (<code>launchpad.exe</code>) - <strong>DO NOT DO THIS ON A PRODUCTION SERVER</strong>.</li>
<li>In both instances of <em>WinDbg</em> enable an <em>Event Filter</em> for <code>C++ EH Exception</code>. If you need to refresh you memory about how to do it, <a href="/2017/04/02/microsoft-sql-server-r-services---internals-ii/">Internals - II</a> goes into the details.</li>
<li>Execute the code in <em>Code Snippet 2</em>.</li>
</ul>

<p>When you execute, you see how you just continue through in the launchpad service, but you break with an exception in SQL Server. It looks like exceptions in the external engine does not affect the launchpad service at all. So let us start looking more in detail what happens in SQL Server.</p>

<h4 id="sql-server-i">SQL Server I</h4>

<p>The exception in SQL server we received above looks like this:</p>

<pre><code class="language-c++">(17d0.3744): C++ EH exception - code e06d7363 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
KERNELBASE!RaiseException+0x68:
00007ffe\`dbba2918 488b8c24c0000000 mov     
rcx,qword ptr [rsp+0C0h] ss:000000e1`817fb9c0=00001f7285a2fbf6
</code></pre>

<p><strong>Code Snippet 5:</strong> <em>External Engine Exception in SQL</em></p>

<p>As in <em>Code Snippet 5</em>, there is an error in SQL Server, and when you continue a second error is raised which you continue through as well. Now execute the code in <em>Code Snippet 2</em> again, and check the call-stack (<code>k</code>) when SQL Server breaks at the first error :</p>

<pre><code class="language-c++">KERNELBASE!RaiseException+0x68
MSVCR120!_CxxThrowException+0xb3 
         [f:\dd\vctools\crt\crtw32\eh\throw.cpp @ 154] 
sqldk!ExceptionBackout::GetCurrentException+0x385
sqldk!ex_raise2+0x52f
sqldk!ex_raise+0xc4
sqlmin!CQScanUdx::GetRow+0x172
sqlmin!CQueryScan::GetRow+0x81
sqllang!CXStmtQuery::ErsqExecuteQuery+0x4dc
...
sqllang!SpExecuteExternalScript+0x140e
...
ntdll!RtlUserThreadStart+0x21
</code></pre>

<p><strong>Code Snippet 6:</strong> <em>Call Stack at Exception</em></p>

<p>Interesting, in <em>Code Snippet 6</em> we see the call to <code>sqlmin!CQScanUdx::GetRow</code> raise an error - even though we are not retrieving any data from the engine! What happens here?</p>

<p>Remember back to the last posts, where we spoke about how the SqlSatellite returns packets to SQL Server after execution. We saw an example of what it looks like if we execute the code in <em>Code Snippet 1</em>:</p>

<p><img src="/images/posts/sql_r_services_19_procmon1.png" alt="" /></p>

<p><strong>Figure 6:</strong> <em>Return Packets</em></p>

<p>We saw <em>Figure 6</em> in <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a>, and we discussed in that post, how the SqlSatellite returns the three packets outlined in red above to SQL Server. The first packet contains metadata information about potential result-sets returned, and if there is no result-set, we still get the packet. The second packet is a packet indicating the end of the result-set, and the third packet can potentially contain values of output parameters. Rows returned for a result-set, appear in packets, between the first and second packet in the figure above.</p>

<p>So, in the case of an error, like in <em>Figure 3</em> above, are there any changes in what the SqlSatellite returns? By now, you probably realise that if I ask a question like that - there is a difference, but let us find out regardless:</p>

<ul>
<li>Run <em>Process Monitor</em> as admin and load the filter we used in <a href="/2017/12/24/microsoft-sql-server-r-services---internals-xvi/">Internals - XVI</a> where we filtered for <code>TCP Send</code> and <code>TCP Receive</code> for <code>BxlServer.exe</code>.</li>
<li>Execute the code in <em>Code Snippet 2</em> (without the <code>Sys.sleep</code> if you want) and look at the output from <em>Process Monitor</em>:</li>
</ul>

<p><img src="/images/posts/sql_r_services_20_procmon1.png" alt="" /></p>

<p><strong>Figure 7:</strong> <em>Process Monitor Output Error</em></p>

<p>It is interesting, that when an error happens, we do not get the packets referred to above. In <em>Figure 7</em> we only see one packet sent back from the SqlSatellite (outlined in red) after the SqlSatellite has received the various packets from SQL Server. While you look at the output from <em>Process Monitor</em>, also make a note of the <code>Path</code> column and the last value (in the figure it is 55538). That value is the port SQL Server receives packets. We use this now in <em>WireShark</em> to try and see if we can get any more information about what is happening.</p>

<p>So, let us switch over to <em>WireShark</em>:</p>

<ul>
<li>Run <em>WireShark</em> as admin.</li>
<li>Choose the network adapter to &ldquo;sniff&rdquo; on. See <a href="/2017/08/29/microsoft-sql-server-r-services---internals-x/">Internals - X</a> for discussion around loop-back adapters etc.</li>
<li>Set a display filter on the port SQL Server listens on (what you saw in the <code>Path</code> column). In this case we want to sniff incoming packets, and if we did it based on what we saw in <em>Figure 7</em> the filter should be: <code>tcp.dstport==55538</code>.</li>
<li>Apply the filter.</li>
</ul>

<p>To start with, we try to see what the packets look like when everything works - and the packets we are interested in, are the two packets with a length of 28. When we execute the code in <em>Code Snippet 1</em> and showing the two data packets as hex, it looks like so:</p>

<pre><code class="language-c++">//first 28 bytes packet
01 00 06 80 1c 00 00 00  b8 16 27 5d 86 b9 95 47
a3 fc 06 0a 17 f1 92 ca  00 00 00 00            

//2:nd 28 bytes packet
01 00 0d 80 1c 00 00 00  b8 16 27 5d 86 b9 95 47
a3 fc 06 0a 17 f1 92 ca  00 00 00 00
</code></pre>

<p><strong>Code Snippet 7:</strong> <em>Packets When No Error</em></p>

<p>When looking at the packets in <em>Code Snippet 7</em>, it looks like the first 4 bytes determines some packet type. Why I say that is when you execute the code multiple times, the first four bytes for packet 1 always look the same, and the same goes for the second packet. The second 4 bytes in both first and second packet, indicates the packet size (<code>1c</code> = 28). I have no &ldquo;clue&rdquo; what the following 16 bytes identify, but it is interesting to notice that the values are the same between packet 1 and 2. The values differ though in between executions. The last four bytes are always 0 in packet 1, in packet two the first byte indicates how many output parameters are returned (if there are any). Remember it is the second packet that contains the output parameters. The last byte in packet 2 is an &ldquo;end&rdquo; byte if parameters are returned, otherwise 0.</p>

<p>So the above is when there are no errors. What when it is? Clear the display in <em>WireShark</em> (easiest way is to stop capturing packets, and start again) and execute the code in <em>Code Snippet 2</em>. As we know from the discussion around <em>Figure 7</em>, we only get one packet back - and when we look at the data as hex it looks like this:</p>

<pre><code class="language-c++">01 00 04 80 1c 00 00 00  31 0b 6c fe b0 9f 7d 43
a4 75 79 5a 70 4c e1 05  12 00 2a 00            
</code></pre>

<p><strong>Code Snippet 8:</strong> <em>Packet When Error</em></p>

<p>The packet in <em>Code Snippet 8</em> is somewhat similar to what we see in <em>Code Snippet 7</em>, in that the first four bytes are always the same, irrespective of the error packet and the second four bytes part indicating the size. What is different and potentially points out that this is an error packet, is the last four bytes. The four last bytes are always the same for any error packet, and as we see - they are not 0.</p>

<p>To summarize what we know so far when an error happens in the external engine:</p>

<ul>
<li>SQL Server receives one 28 bytes data packet back from the SqlSatellite.</li>
<li>The four last bytes always are <code>12 00 2a 00</code>.</li>
</ul>

<p>Let&rsquo;s go back to the call-stack we see in <em>Code Snippet 6</em> and try to figure out what happens in the code. We have already said the error happens in <code>sqlmin!CQScanUdx::GetRow</code>. We have not spoken much about <code>GetRow</code> at all, but if you remember in <a href="/2018/01/03/microsoft-sql-server-r-services---internals-xvii/">Internals- XVII</a> we discussed about how <code>sqllang!CXStmtQuery::ErsqExecuteQuery</code> called <code>sqllang!CUDXR_ExternalScript::PullRow</code>. In fact <code>sqllang!CXStmtQuery::ErsqExecuteQuery</code> calls <code>sqlmin!CQueryScan::GetRow</code> which in turn calls <code>sqlmin!CQScanUdx::GetRow</code> and it calls <code>PullRow</code>.</p>

<p>In <a href="/2018/01/03/microsoft-sql-server-r-services---internals-xvii/">Internals - XVII</a> we traced the <code>PullRow</code> call and saw how SQL Server handled returning data. To see what happens during an error, let us do the trace one step up - in <code>sqlmin!CQScanUdx::GetRow</code>. We start with the code that works (<em>Code Snippet 1</em>):</p>

<ul>
<li>Change the <code>Sys.sleep</code> in the script in *Code Snippet 1` to 30 seconds.</li>
<li>Set a break-point in the SQL Server process at <code>sqlmin!CQScanUdx::GetRow</code>.</li>
<li>Execute the code in <em>Code Snippet 1</em>.</li>
<li>When <code>sqlmin!CQScanUdx::GetRow</code> hits the breakpoint, do a <code>wt -l8</code> (trace and watch the execution 8 levels deep).</li>
</ul>

<p>Look at the <em>WinDbg</em> output when the `Sys.sleep(30) is hit in the script, and notice the last few lines look like so:</p>

<pre><code class="language-c++">sqllang!EventInternal&lt;SuspendQueueSLock&gt;::Wait
 KERNEL32!TlsGetValueStub
 KERNEL32!TlsGetValue
sqllang!EventInternal&lt;SuspendQueueSLock&gt;::Wait
 sqldk!SystemThread::GetCurrentId
sqllang!EventInternal&lt;SuspendQueueSLock&gt;::Wait
 sqldk!SOS_Task::PreWait
sqllang!EventInternal&lt;SuspendQueueSLock&gt;::Wait
 sqldk!SOS_Scheduler::Suspend
</code></pre>

<p><strong>Code Snippet 9:</strong> <em>wt Output at Sys.sleep</em></p>

<p>The code in <em>Code Snippet 9</em> is what precedes the code we saw in <a href="/2018/01/03/microsoft-sql-server-r-services---internals-xvii/">Internals - XVII</a> where the execution waited for a signal. After the trace has finished, copy it to a text file to make it easier to investigate what happens. Then execute the code in <em>Code Snippet 2</em> and do a new trace, and copy that result into a new file as well. When you then compare the two files you see how the first part is identical:</p>

<pre><code class="language-c++">sqlmin!CQScanUdx::GetRow
  sqllang!CUDXR_ExternalScript::PullRow
    sqllang!CUDXR_ExternalScript::ProcessOutputSchema
      sqllang!CUDXR_ExternalScript::GetOutputSchema
        sqllang!CServerCargoContext::ReadSchemaInternal
          sqllang!CSQLSatelliteCommunication::CSQLSatelliteCommunication
          ...
          sqllang!CServerCargoContext::ReadAllPackets
            sqllang!CSQLSatelliteConnection::GetNextMessage
              ...
              sqllang!EventInternal&lt;SuspendQueueSLock&gt;::Wait
                sqldk!SOS_Scheduler::Suspend
              //here is Sys.sleep

              //here we continue
                sqldk!SOS_Scheduler::SuspendNonPreemptive
              sqllang!EventInternal&lt;SuspendQueueSLock&gt;::Wait
               ...
            sqllang!CSQLSatelliteConnection::GetNextMessage
              sqllang!CSQLSatelliteConnection::RetrieveMessage
                ...
          sqllang!CServerCargoContext::ReadAllPackets
            sqllang!CSQLSatelliteCommunication::BindReadSNIPacket
              sqllang!CSQLSatelliteCommunication::GetSQLSatelliteMessage
              ...
            ...
          sqllang!CServerCargoContext::ReadAllPackets
          ...
</code></pre>

<p><strong>Code Snippet 10:</strong> <em>Common Code Path</em></p>

<p>In <em>Code Snippet 10</em> we recognise from <a href="/2018/01/03/microsoft-sql-server-r-services---internals-xvii/">Internals - XVII</a> the calls to retrieve rows, and they are the same in both trace files up until around line 50, after the <code>ReadAllPackets</code> in the next to last line above. Continuing,  what we see in the files change. In the no error trace file nothing out of the ordinary happens, and eventually, the trace ends with the call to <code>sqllang!CUDXR_ExternalScript::ProcessOutputArguments</code>, which we discussed in <a href="/2018/01/10/microsoft-sql-server-r-services---internals-xviii/">Internals - XVIII</a>. The error trace file looks completely different though from after the last <code>ReadAllPackets</code> above:</p>

<pre><code class="language-c++">sqlmin!CQScanUdx::GetRow
  sqllang!CUDXR_ExternalScript::PullRow
    sqllang!CUDXR_ExternalScript::ProcessOutputSchema
      sqllang!CUDXR_ExternalScript::GetOutputSchema
        sqllang!CServerCargoContext::ReadSchemaInternal
        ...
          
          sqllang!CServerCargoContext::ReadAllPackets
            sqllang!CSQLSatelliteCommunication::BindReadSNIPacket
              sqllang!CSQLSatelliteCommunication::GetSQLSatelliteMessage 
              ...
          //this is the last ReadAllPackets above
           sqllang!CServerCargoContext::ReadAllPackets
            sqllang!CServerCargoContext::HandleSpecialMessage
              sqllang!CServerCargoContext::HandleAbortMessage
              ...
            ...
    sqllang!CUDXR_ExternalScript::ProcessOutputSchema
  sqllang!CUDXR_ExternalScript::PullRow
sqlmin!CQScanUdx::GetRow
  sqllang!CUDXC_ExternalScript::EUdxTypeGet
sqlmin!CQScanUdx::GetRow
  ...
...
  sqldk!ex_raise
    sqldk!ex_raise2
    ...
           ntdll!RtlRaiseException
           ...
</code></pre>

<p><strong>Code Snippet 11:</strong> <em>Code Path for Exception</em></p>

<p>Included in <em>Code Snippet 11</em> is the start of the routine, the same as in <em>Code Snippet 10</em>, but we see how after the last <code>ReadAllPackets</code> the call goes into <code>sqllang!CServerCargoContext::HandleSpecialMessage</code> and <code>sqllang!CServerCargoContext::HandleAbortMessage</code>. These two calls are still inside the <code>PullRow</code> and <code>ProcessOutputSchema</code> calls, e.g. while SQL Server processes the packet from the SqlSatellite. The <code>HandleSpecialMessage</code> and <code>HandleAbortMessage</code> explains the packet in <em>Code Snippet 8</em>: it is an abort packet.</p>

<p>Further down in the code in <em>Code Snippet 11</em> is a call to <code>sqllang!CUDXC_ExternalScript::EUdxTypeGet</code>, where - I believe - SQL Server retrieves the type of error and then raises the error in the <code>sqldk!ex_raise</code> call-chain. The question now is how SQL Server knows the error message? What I mean is: when we look at the <code>text</code> column in the result from <em>Code Snippet 4</em> we see for error number the text being: <code>An external script error occurred: %.*ls%.*ls</code>. The <code>%.*ls%.*ls</code> characters are substitution parameters and when we look at the actual error message in <em>Figure 3</em>, the replacement for the substitution parameters is what we see outlined in red below:</p>

<p><img src="/images/posts/sql_r_services_20_error3.png" alt="" /></p>

<p><strong>Figure 8:</strong> <em>Substitution Text</em></p>

<p>So, where do the substitution outlined in <em>Figure 8</em> come from, and what is the text outlined in blue in the same figure? We know that the actual error-text is not part of the 28 bytes packet from the SqlSatellite, as no matter how long the error message is, the length of the packet is always the same. So where have we seen messages being passed back to SQL Server, without being in any data packets from the SqlSatellite? Well, we saw it in <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a> where the launchpad service did send &ldquo;print&rdquo; messages as well as ack messages from the external engine to SQL Server. So, let us go to the launchpad service and <a href="http://queue.acm.org/detail.cfm?id=945136">&ldquo;spelunk&rdquo;</a> around a bit:</p>

<h4 id="launchpad">Launchpad</h4>

<p>We know, from <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a>, that the two routines used when sending messages are:</p>

<ul>
<li><code>launchpad!CSatelliteCargoContext::SendLogMessage</code></li>
<li><code>launchpad!CSQLSatelliteCommunication::SendAckMessage</code></li>
</ul>

<p>First, we try to find any routines handling error messages specifically: <code>x *!*Send*Error*</code>. That returns one routine: <code>launchpad!HostErrorReportingManager:: SendErrorToUserWithTracing</code>. I did however not get a hit at a break-point set at the routine, when I executed the code in <em>Code Snippet 2</em>. So it looks like our best bet is one of the previously mentioned routines, and we try <code>launchpad!CSatelliteCargoContext::SendLogMessage</code> first:</p>

<ul>
<li>Set a break-point <code>launchpad!CSatelliteCargoContext::SendLogMessage</code>.</li>
<li>Execute the code in <em>Code Snippet 1</em> (no error code).</li>
</ul>

<p>You see how the code executes all the way through, and the breakpoint is not hit. However, the execution breaks at the break-point when you run  <em>Code Snippet 2</em> (error)! The only difference between the two code snippets is the error in <em>Code Snippet 2</em>, so it looks like the launchpad service sends a message to SQL Server with the error text from the external engine.</p>

<h4 id="sql-server-ii">SQL Server II</h4>

<p>We know from <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a> that <code>sqllang!CUDXR_ExternalScript::HandleLogMessage</code> handles messages that are sent through <code>launchpad!CSatelliteCargoContext::SendLogMessage</code>. I wonder if that is the case when there is an exception as well:</p>

<ul>
<li>Disable the break-points in the launchpad <em>WinDbg</em> instance.</li>
<li>Disable any break-points in the SQL Server <em>WinDbg</em> instance.</li>
<li>Set a break-point at <code>sqllang!CUDXR_ExternalScript::HandleLogMessage</code>.</li>
<li>Run the <em>Code Snippet 2</em> code.</li>
<li>Continue through the exceptions.</li>
</ul>

<p>Cool, we hit the <code>HandleLogMessage</code> break-point, so let us look at the call-stack (<code>k</code>):</p>

<pre><code class="language-c++">sqllang!CUDXR_ExternalScript::HandleLogMessage
sqllang!CSQLSatelliteCommunication::WaitForAckMessage+0x164
sqllang!CUDXR_ExternalScript::ShutdownService+0x250
sqlmin!CUDXR_ExternalScriptCleanup::Close+0xe
sqlmin!CQueryExecContext::FinalCleanupNoX+0x116
sqlmin!CQueryScan::DestroyQueryOnException+0x129
sqllang!CXStmtQuery::ShutdownOnException+0xc8
sqllang!CXStmtQuery::FinishOnExceptionImp+0x67
sqllang!_isa_available_init+0xe8cf3
MSVCR120!_CallSettingFrame+0x20 
               [f:\dd\vctools\crt\crtw32\eh\amd64\handlers.asm @ 51] 
MSVCR120!__CxxCallCatchBlock+0xf5 
               [f:\dd\vctools\crt\crtw32\eh\frame.cpp @ 1281] 
ntdll!RcConsolidateFrames+0x3
sqllang!CMsqlExecContext::FExecute+0xa9e
sqllang!CSQLSource::Execute+0x983
sqllang!SpExecuteExternalScript+0x140e
...
ntdll!RtlUserThreadStart+0x21
</code></pre>

<p><strong>Code Snippet 12:</strong> <em>HandleLogMessage Call Stack</em></p>

<p>The call-stack we see in <em>Code Snippet 12</em> indicates a similar code path for <code>HandleLogMessage</code> as we saw in <em>Code Snippet 12</em> in the <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a> post . They are similar, but there are some differences which becomes clear when we compare the code paths between the two call-stacks. In a no error scenario we have a code path like this:</p>

<pre><code class="language-c++">sqllang!CMsqlExecContext::FExecute
  sqllang!CMsqlExecContext::ExecuteStmts&lt;1,1&gt;
    sqllang!CXStmtSelect::XretExecute
      sqllang!CXStmtQuery::ErsqExecuteQuery
        sqllang!CXStmtQuery::ShutdownNormal
          sqlmin!CQueryScan::DestroyQuery
            sqlmin!CQueryExecContext::FinalCleanup
              sqlmin!CUDXR_ExternalScriptCleanup::Close
              ...
              ...
                sqllang!CUDXR_ExternalScript::HandleLogMessage

</code></pre>

<p><strong>Code Snippet 13:</strong> <em>Code Path HandleLogMessage No Error</em></p>

<p>Whereas the code path for an exception looks like so:</p>

<pre><code class="language-c++">sqllang!CMsqlExecContext::FExecute
  ...
  ...
  sqllang!CXStmtQuery::FinishOnExceptionImp
    sqllang!CXStmtQuery::ShutdownOnException
      sqlmin!CQueryScan::DestroyQueryOnException
        sqlmin!CQueryExecContext::FinalCleanupNoX
          sqlmin!CUDXR_ExternalScriptCleanup::Close
            ...
            ...
              sqllang!CUDXR_ExternalScript::HandleLogMessage
</code></pre>

<p><strong>Code Snippet 14:</strong> <em>Code Path HandleLogMessage with Error</em></p>

<p>When we compare <em>Code Snippet 13</em> and <em>Code Snippet 14</em> we see how - in both cases - the code goes into shutdown and cleanup routines, but in the error case these routines are &ldquo;exception&rdquo; routines:</p>

<ul>
<li><code>sqllang!CXStmtQuery::FinishOnExceptionImp</code></li>
<li><code>sqllang!CXStmtQuery::ShutdownOnException</code></li>
<li><code>sqlmin!CQueryScan::DestroyQueryOnException</code></li>
</ul>

<p>The question now is how the message received from the external engine via the launchpad service is &ldquo;hooked&rdquo; up to the actual exception. The assumption is that the answer to that has something to do with <code>HandleLogMessage</code>. So, let us compare what happens in <code>HandleLogMessage</code> when we run code that logs a message (as we did in <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a>), with running code that causes an exception.</p>

<p>This is the code with no exception:</p>

<pre><code class="language-sql">EXEC sp_execute_external_script
                 @language = N'R' ,
                 @script = N'Sys.sleep(10)
                             cat(&quot;Hello World&quot;)
                             d &lt;- 42';
</code></pre>

<p><strong>Code Snippet 15:</strong> <em>Code with Print</em></p>

<p>The code in <em>Code Snippet 15</em> is the same code we used in <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Internals - XIX</a>, and this is what we do:</p>

<ul>
<li>Set a breakpoint at <code>sqllang!CUDXR_ExternalScript::HandleLogMessage</code>.</li>
<li>Run the code in <em>Code Snippet 15</em>.</li>
<li>Do a trace, (<code>wt -l6</code>), of the <code>HandleLogMessage</code> call when you hit the breakpoint.</li>
<li>Copy the output of the trace into a new text file.</li>
<li>Make sure the <code>C++ EH Exception</code> event filter is still enabled.</li>
<li>Run the code in <em>Code Snippet 2</em>.</li>
<li>Continue through the exceptions.</li>
<li>Do a trace, (<code>wt -l6</code>), of the <code>HandleLogMessage</code> call when you hit the breakpoint.</li>
<li>Copy the output of the trace into a new text file.</li>
</ul>

<p>When you compare the two files you see they are similar, but at around line 190, or so, there is a difference. In the &ldquo;no-error&rdquo; file you see something like so:</p>

<pre><code class="language-c++">sqllang!CErrorReportingManager::CwchFormatAndPrint
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
sqllang!CErrorReportingManager::CwchFormatAndPrint
  sqllang!CErrorReportingManager::SetError
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!CErrorReportingManager::SetError
  sqllang!getret
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!getret
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!getret
sqllang!CErrorReportingManager::CwchFormatAndPrint
  sqllang!CErrorReportingManager::SendErrorToUser
  ...
  sqllang!ErrorReportedAutoPublish::Publish
</code></pre>

<p><strong>Code Snippet 16:</strong> &ldquo;HandleLogMessage No Error&rdquo;</p>

<p>However in the &ldquo;error&rdquo; file you see:</p>

<pre><code class="language-c++">sqllang!CErrorReportingManager::CwchFormatAndPrint
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
    sqllang!PCXCtxtGetIfExistsEXPENSIVE
      KERNEL32!TlsGetValueStub
      KERNEL32!TlsGetValue
    sqllang!PCXCtxtGetIfExistsEXPENSIVE
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
    sqllang!PCXCtxtGetIfExistsEXPENSIVE
      KERNEL32!TlsGetValueStub
      KERNEL32!TlsGetValue
    sqllang!PCXCtxtGetIfExistsEXPENSIVE
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
    sqllang!CExecLvlRepresentation::FErrorModeJumpToCatch
  sqllang!CAutoInterceptErrorForTsqlCatch::Initialize
sqllang!CErrorReportingManager::CwchFormatAndPrint
  sqllang!CErrorReportingManager::SetError
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!CErrorReportingManager::SetError
  sqllang!getret
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!getret
    KERNEL32!TlsGetValueStub
    KERNEL32!TlsGetValue
  sqllang!getret
sqllang!CErrorReportingManager::CwchFormatAndPrint
  sqllang!CErrorReportingManager::SendErrorToUser
  ...
  sqllang!ErrorReportedAutoPublish::Publish
</code></pre>

<p><strong>Code Snippet 17:</strong> &ldquo;HandleLogMessage Exception&rdquo;</p>

<p>You can see how, in the &ldquo;error&rdquo; file (<em>Code Snippet 17</em>), the code loops on <code>sqllang!CAutoInterceptErrorForTsqlCatch::Initialize</code>, before it calls into <code>sqllang!CExecLvlRepresentation::FErrorModeJumpToCatch</code>. This loop does not happen in the &ldquo;non-error&rdquo; case. Oh, and it is interesting to see how in the case of &ldquo;normal&rdquo; log messages, the message is still considered being an error:</p>

<ul>
<li><code>sqllang!CErrorReportingManager::SetError</code></li>
<li><code>sqllang!CErrorReportingManager::SendErrorToUser</code></li>
<li><code>sqllang!ErrorReportedAutoPublish::Publish</code></li>
</ul>

<h2 id="errors-unknown-to-sql-server">Errors Unknown to SQL Server</h2>

<p>This post has become long, but let us talk a little bit about the type of error we saw in <em>Figure 4</em> before we finish. I refer to where SQL Server does not know about the error in the external engine (you see no specific external engine SQL Server error message).</p>

<p>If we look at how the error in <em>Code Snippet 2</em> is handled and compare it with the error in <em>Code Snippet 3</em>, we may think they are handled the same way (an error is an error, right?), and it may look like that. For example, if we look at <em>Process Monitor</em> we see the same 28 bytes packet sent to SQL Server as we saw in <em>Figure 7</em>. The packet itself looks the same as in <em>Code Snippet 8</em> (first 4 bytes - <code>01 00 04 80</code> second 4 bytes - <code>1c 00 00 00</code>, and the final 4 bytes - <code>12 00 2a 00</code>). These similarities led me to believe we would see the same flow as in <em>Code Snippet 11</em>, where the code goes into <code>sqllang!CServerCargoContext::HandleSpecialMessage</code> and further down into <code>sqllang!CServerCargoContext::HandleAbortMessage</code>. I was wrong!</p>

<p>I realised the errors of my ways when I set a breakpoint at <code>sqllang!CServerCargoContext::HandleSpecialMessage</code>, ran the code in <em>Code Snippet 3</em>, and the code never hit the breakpoint! So what happened here? Well, the actual exception happened way before sending the <code>@input_data_1</code> data to the SqlSatellite, and that is the cause of what we see. So let us:</p>

<ul>
<li>Disable all breakpoints.</li>
<li>Keep the <code>C++ EH Exception</code> event filter enabled.</li>
</ul>

<p>When we run the <em>Code Snippet 3</em> code, we see how the execution breaks at exception three times in quick succession. When inspecting the call-stack of each of those three occurrences we see how all three exceptions are related to the <code>sqlTsEs!ESArithErrorHandler</code> routine which handles, among other things, division by zero.</p>

<p>There is a slight pause when continuing after the third exception, and then we break once again with a new exception. The call-stack now looks like so:</p>

<pre><code class="language-c++">KERNELBASE!RaiseException+0x68
MSVCR120!_CxxThrowException+0xb3 
          [f:\dd\vctools\crt\crtw32\eh\throw.cpp @ 154] 
sqldk!ExceptionBackout::GetCurrentException+0x385
sqldk!ex_raise2+0x52f
sqldk!ex_raise+0xc4
sqlTsEs!ESArithErrorHandler+0x159
sqlTsEs!CTEsArith&lt;3,56,56,1,0&gt;::RsltArithArgArg+0x94
sqlTsEs!CEsExec::GeneralEval4+0xe7
sqllang!CUDXR_ExternalScript::PushRow+0x78
sqlmin!CQScanUdx::PushNextChildRow+0x13b
sqlmin!CQScanUdx::GetRow+0x79
sqlmin!CQueryScan::GetRow+0x81
sqllang!CXStmtQuery::ErsqExecuteQuery+0x4dc
...
sqllang!SpExecuteExternalScript+0x140e
...
ntdll!RtlUserThreadStart+0x21
</code></pre>

<p><strong>Code Snippet 18:</strong> &ldquo;Exception after Division by Zero&rdquo;</p>

<p>The call-stack we see in <em>Code Snippet 18</em> shows how an exception happens when we try to push data rows to a packet to send to SqlSatellite. So, even though there is a division by zero exception initially, SQL Server still tries to push data to the satellite. The question is if some data goes to the satellite? Remember from previous posts how we covered data sent to the SqlSatellite and how the satellite received end-packet(s) after the data. I wonder what happens in this case:</p>

<ul>
<li>Disable the <code>C++ EH Exception</code> event filter.</li>
<li>Run <em>Process Monitor</em> as admin, and use the event filter we used above.</li>
<li>Execute the code in <em>Code Snippet 3</em> and look at the output from <em>Process Monitor</em>:</li>
</ul>

<p><img src="/images/posts/sql_r_services_20_procmon2.png" alt="" /></p>

<p><strong>Figure 9:</strong> <em>Process Monitor Output 2</em></p>

<p>When we compare <em>Figure 9</em> with what we see in <em>Figure 7</em>, there is not much of a difference. In both figures we see SQL Server sending a 28 bytes end-packet to the SqlSatellite. In <em>Figure 9</em> it is outlined in red, and in <em>Figure 7</em> it is just above the highlighted packet. Even though the two figures look the same, the difference lies in the end-packet. I used <em>WireShark</em> to capture packets from <em>Code Snippet 1</em> (no error), <em>Code Snippet 2</em> (external engine error) and <em>Code Snippet 3</em> (SQL error):</p>

<pre><code class="language-c++">//OK
01 00 06 80 1c 00 00 00  76 01 1b 35 fa 75 64 48
81 34 e2 ec ec 7d 2d c6  00 00 00 00   

//engine exception
01 00 06 80 1c 00 00 00  90 2b 8a 3f 60 5c d5 49
a0 8f f0 c4 c0 bf c0 cb  00 00 00 00            

//exception SQL
01 00 04 80 1c 00 00 00  7f 82 bc e0 3f 5a 29 48
83 d7 20 46 ad 79 04 21  00 00 00 00  
</code></pre>

<p><strong>Code Snippet 19:</strong> <em>Different End Packet Types</em></p>

<p>When we look at the packets in <em>Code Snippet 19</em> we see how the two first packets look like the success packets SQL Server receives from the SqlSatellite. Why I say that, is because the four first bytes are the same: <code>01 00 06 80</code>. The third packet, however, looks like the &ldquo;error&rdquo; packet SqlSatellite sends to SQL Server, with a header of <code>01 00 04 80</code>. The difference between the packet that SQL Server sends to the satellite, and the packet SQL Server receives are the last four bytes. Sending to the satellite the bytes are <code>00 00 00 00</code>, whereas when an error packet comes back from the satellite, it is - as we discussed when we saw <em>Code Snippet 8</em>  - <code>12 00 2a 00</code>. I do believe that this packet is what causes the SqlSatellite to raise the error we see <code>SqlSatellite cannot read data chunk ...</code>.</p>

<p>So this is what happens:</p>

<ul>
<li>The exception happens in the SQL Server engine, before the process has been handed over to the SqlSatellite.</li>
<li>An error packet is sent to the satellite.</li>
<li>The satellite cannot process the packet.</li>
<li>The SqlSatellite sends back the error message we see.</li>
<li>This is handled by the <code>HandleLogMessage</code> routine in SqlServer.</li>
</ul>

<p>In the bullet points above I say that the <code>HandleLogMessage</code> routine handles the message, and you can see this by setting a breakpoint on this routine.</p>

<h2 id="summary">Summary</h2>

<p>We have in this post seen how there are two types of errors related to the external engine:</p>

<ul>
<li>Error raised in the extern engine itself.</li>
<li>Error happening before the external engine has taken over.</li>
</ul>

<p>When an error happens in the external engine:</p>

<ul>
<li>the SqlSatellite sends an error packet to SQL Server.</li>
<li>SQL Server receives the packet and handles it in <code>sqllang!CServerCargoContext::HandleSpecialMessage</code> and <code>sqllang!CServerCargoContext::HandleAbortMessage</code>. This raises a SQL Server exception related to the external engine.</li>
<li>The error message from the external engine is sent to the launchpad service.</li>
<li>The launchpad service sends the message to SQL Server via <code>launchpad!CSatelliteCargoContext::SendLogMessage</code></li>
<li>SQL Server handles the message in <code>sqllang!CUDXR_ExternalScript::HandleLogMessage</code>.</li>
</ul>

<p>When the error happens in SQL Server before the external engine:</p>

<ul>
<li>An error packet is sent to the satellite.</li>
<li>The satellite cannot process the packet.</li>
<li>The SqlSatellite sends back the error message via <code>launchpad!CSatelliteCargoContext::SendLogMessage</code>.</li>
<li>SQL Server handles the message in <code>sqllang!CUDXR_ExternalScript::HandleLogMessage</code> and related routines.</li>
</ul>

<h2 id="finally">~ Finally</h2>

<p>If you have comments, questions etc., please comment on this post or <a href="mailto:niels.it.berglund@gmail.com">ping</a> me.</p>
    <br/>
    
<h2>Blog Feed:</h2>
To automatically receive more posts like this, please
<a href="http://feeds.feedburner.com/manageddata/" target="_blank"> subscribe to my RSS/Atom feed</a> in your feed reader!</p>



  </div>
  <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>

<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-3005153158271538"
     data-ad-slot="1158080725"
     data-ad-format="auto"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>
  

<div class="post--navigation post--navigation-single">
    
    <a href="/2018/01/28/interesting-stuff---week-4/" class="post--navigation-prev">
      <i aria-hidden="true" class="fa fa-chevron-left"></i>
      <span class="navigation-tittle">Interesting Stuff - Week 4</span>
    </a>
    
    
    <a href="/2018/02/04/interesting-stuff---week-5/" class="post--navigation-next">
      <span class="navigation-tittle">Interesting Stuff - Week 5</span>
      <i aria-hidden="true" class="fa fa-chevron-right"></i>
    </a>
    
</div>


  


<div class="post__related">
    
    <h2>Related Articles</h2>
    <ul class="related-posts">
        
<li>
  <span class="list__title--small">
    <a href="/2018/01/20/microsoft-sql-server-r-services---internals-xix/">Microsoft SQL Server R Services - Internals XIX</a>
      
      <time class="pull-right hidden-tablet">Jan 20 &#39;18</time>
      
  </span>
</li>

<li>
  <span class="list__title--small">
    <a href="/2018/01/10/microsoft-sql-server-r-services---internals-xviii/">Microsoft SQL Server R Services - Internals XVIII</a>
      
      <time class="pull-right hidden-tablet">Jan 10 &#39;18</time>
      
  </span>
</li>

<li>
  <span class="list__title--small">
    <a href="/2018/01/03/microsoft-sql-server-r-services---internals-xvii/">Microsoft SQL Server R Services - Internals XVII</a>
      
      <time class="pull-right hidden-tablet">Jan 03 &#39;18</time>
      
  </span>
</li>

<li>
  <span class="list__title--small">
    <a href="/2017/08/29/microsoft-sql-server-r-services---internals-x/">Microsoft SQL Server R Services - Internals X</a>
      
      <time class="pull-right hidden-tablet">Aug 29 &#39;17</time>
      
  </span>
</li>

<li>
  <span class="list__title--small">
    <a href="/2017/12/24/microsoft-sql-server-r-services---internals-xvi/">Microsoft SQL Server R Services - Internals XVI</a>
      
      <time class="pull-right hidden-tablet">Dec 24 &#39;17</time>
      
  </span>
</li>

    </ul>
</div>



  
    
        <div id="disqus_thread"></div>
<script type="text/javascript">
    

    (function () {
    if (location.hostname === "localhost" ||
      location.hostname === "127.0.0.1" ||
      location.hostname === "") {
      return;
    }
    var dsq = document.createElement('script');
    dsq.type = 'text/javascript';
    dsq.async = true;
    var disqus_shortname = 'manageddata';
    dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
    (document.getElementsByTagName('head')[0] || 
      document.getElementsByTagName('body')[0]).appendChild(dsq);
  })();
</script>

<noscript>
  Please enable JavaScript to view the
  <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a>
</noscript>
<a href="http://disqus.com/" class="dsq-brlink">comments powered by
  <span class="logo-disqus">Disqus</span>
</a>

    


</article>


        </div>
        
    
  
  
<script>
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-18914734-2', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>


<script defer src="https://use.fontawesome.com/releases/v5.2.0/js/all.js"
  integrity="sha384-4oV5EgaV02iISL2ban6c/RmotsABqE4yZxZLcYMAdG7FAPsyHYAPpywE9PJo+Khy"
  crossorigin="anonymous">
</script>


<script src="/js/highlight.pack.js"></script>

<script type="text/javascript">
    hljs.configure({languages: []});
    hljs.initHighlightingOnLoad();
</script>



    



    </body>
</html>
